docs/source/rfc/on-chain-scripting.org

```diff
@@ -0,0 +1,954 @@
+
+Firstly, owing to the maturity and the ubiquity of the standard much
+work has gone towards mitigating the performance issues associated to
+garbage collection in Java.  [[https://plummerssoftwarellc.github.io/PrimeView/report?id=3990&hi=False&hf=False&hp=False&fi=&fp=&fa=&ff=&fb=&tp=False&sc=pp&sd=True][The top-end procedural Java programs can
+compete with mid-level C and C++ programs]].  [[https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/design.html#compiling_loading_and_linking_native_methods][Since Java 8]], it is
+possible to link dynamically and statically against native methods:
+while this comes with some additional overhead, it would allow C-ABI
+to work as a /lingua franca/ but idiomatic error handling via
+exceptions would have to be implemented as an extension. This also
+answers the question of Rust interoperability: it is possible to do so
+by exposing the FFI (the added benefit being that the FFI could be
+later used for a different standard, and would have to be implemented
+for WASM anyway).  The =iroha-java= library makes use of the Parity
+SCALE codec.  This is the first instance where the library is not the
+reference implementation so it should be noted that there is
+potential for bugs. Fortunately, the SCALE standard is well-specified
+and the problem is mitigated.
+
+Java offers memory safety via garbage collection.  Java does not have
+special handling for foreign functions and thus it is possible that
+the aforementioned functions can themselves cause undefined behaviour.
+This is a step down from the guarantees of Rust, and it is possible
+that standard and well-tested [[https://www.cisa.gov/news-events/news/apache-log4j-vulnerability-guidance][Java applications can still have
+vulnerabilities]].  Not to say that Java applications are insecure,
+but the percentage is slightly higher than for Rust applications, to
+have a CVE, due to a focus on simplicity of syntax obscuring potential
+points of undefined behaviour.
+
+Java binaries need to be bench marked in a representative form to be
+comparable.  There are some architectural reasons to suspect that Java
+binaries can be smaller than equivalent statically linked Rust
+binaries: mainly because most of the standard library is linked
+dynamically by the Java Runtime Environment, so one gets most of the
+benefit of the dynamic linkage already.  However most objects in the
+Iroha data model are defined as Rust structures, so interactions with
+them must be done via either intermediate representations, or dynamic
+linkage. As such dynamic linkage is a must.  Fortunately, it is easier
+to do that with Java than it is with Rust, even though this is at the
+cost of performance, most dispatch in Java is dynamic by design, thus
+making the question of handling object lifetimes inside libraries much
+easier.  Dynamic linkage of some form, thus becomes a necessity.
+
+Java is opaque: it was designed for the open-closed principle, where
```

Comment:
> Java is opaque

I would argue that JVM code (if not obfuscated) is much easier to inspect than x86/wasm, as the compilation preserves more structure of the program (along with class/function names!). A lot of times, a good decompiler can easily recover the equivalent java source code